src/lib/yang/path_data.lua

```diff
@@ -431,6 +431,81 @@ function remover_for_schema_by_name (schema_name, path)
 end
 remover_for_schema_by_name = util.memoize(remover_for_schema_by_name)
 
+function consistency_checker_from_grammar(grammar)
+   -- Converts a relative path to an absolute path.
+   local function to_absolute_path (leafref, path)
+      if leafref:sub(1, 2) == './' then
+         leafref = leafref:sub(3)
+         return path..'/'..leafref
+      end
+      while leafref:sub(1, 3) == '../' do
+         leafref = leafref:sub(4)
+         path = lib.dirname(path)
+      end
+      return path..'/'..leafref
+   end
+   local function leafref (node)
+      return node.argument_type and node.argument_type.leafref
+   end
+   -- Leafref nodes iterator. Returns node as value and full data path as key.
+   local function visit_leafref_paths (root)
+      local function visit (path, node)
+         if node.type == 'struct' then
+            for k,v in pairs(node.members) do visit(path..'/'..k, v) end
+         elseif node.type == 'array' then
+            -- Pass.
+         elseif node.type == 'scalar' then
+            if leafref(node) then
+               coroutine.yield(path, node)
+            else
+               -- Pass.
+            end
+         elseif node.type == 'table' then
+            for k,v in pairs(node.keys) do visit(path..'/'..k, v) end
+            for k,v in pairs(node.values) do visit(path..'/'..k, v) end
+         elseif node.type == 'choice' then
+            for _,choice in pairs(node.choices) do
+               for k,v in pairs(choice) do visit(path..'/'..k, v) end
+            end
+         else
+            error('unexpected kind', node.kind)
+         end
+      end
+      return coroutine.wrap(function() visit('', root) end), true
+   end
+   -- Fetch value of path in data tree.
+   local function resolve (data, path)
+      local ret = data
+      for k in path:gmatch("[^/]+") do ret = ret[k] end
+      return ret
+   end
+   -- If not present, should be true.
+   local function require_instance (node)
+      if node.argument_type.require_instances == nil then return true end
+      return node.argument_type.require_instances
+   end
+   return function (data)
+      for path, node in visit_leafref_paths(grammar) do
```

Comment:
I think we want to collect the "interesting" path-node pairs outside the returned consistency checker -- it allows more precomputation.  We might be able to pre-resolve the getter as well!